runtime/compiler/env/J9KnownObjectTable.cpp

```diff
@@ -56,28 +65,167 @@ J9::KnownObjectTable::isNull(Index index)
-   _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+#if defined(JITSERVER_SUPPORT)
+   if (self()->comp()->isOutOfProcessCompilation())
+      {
+      auto stream = TR::CompilationInfo::getStream();
+      stream->write(JITServer::MessageType::KnownObjectTable_getIndex, objectPointer);
+      auto recv = stream->read<TR::KnownObjectTable::Index, uintptrj_t *>();
+
+      TR::KnownObjectTable::Index index = std::get<0>(recv);
+      if (index != nextIndex)
+         {
+         return index;
+         }
+      else
+         {
+         _references.setSize(nextIndex+1);
+         _references[nextIndex] = std::get<1>(recv);
+         }
+      }
+   else
+#endif /* defined(JITSERVER_SUPPORT) */
+      {
+      TR_J9VMBase *fej9 = (TR_J9VMBase *)(self()->fe());
+      TR_ASSERT(fej9->haveAccess(), "Must haveAccess in J9::KnownObjectTable::getIndex");
+
+      // Search for existing matching entry
+      //
+      for (uint32_t i = 1; i < nextIndex; i++)
+         if (*_references.element(i) == objectPointer)
+            return i;
+
+      // No luck -- allocate a new one
+      //
+      J9VMThread *thread = getJ9VMThreadFromTR_VM(self()->fe());
+      TR_ASSERT(thread, "assertion failure");
+      _references.setSize(nextIndex+1);
+      _references[nextIndex] = (uintptrj_t*)thread->javaVM->internalVMFunctions->j9jni_createLocalRef((JNIEnv*)thread, (j9object_t)objectPointer);
+      }
+
    return nextIndex;
    }
 
 
+TR::KnownObjectTable::Index
+J9::KnownObjectTable::getIndex(uintptrj_t objectPointer, bool isArrayWithConstantElements)
+   {
+   TR::KnownObjectTable::Index index = self()->getIndex(objectPointer);
+   if (isArrayWithConstantElements)
+      {
+      self()->addArrayWithConstantElements(index);
+      }
+   return index;
+   }
+
+
+uintptrj_t
+J9::KnownObjectTable::dereferenceObjectPointerReference(uintptrj_t *objectReferenceLocation)
+   {
+   uintptrj_t objectPointer = 0;
+#if defined(JITSERVER_SUPPORT)
+   if (self()->comp()->isOutOfProcessCompilation())
+      {
+      auto stream = TR::CompilationInfo::getStream();
+      stream->write(JITServer::MessageType::KnownObjectTable_dereferenceObjectPointerReference, objectReferenceLocation);
+      objectPointer = std::get<0>(stream->read<uintptrj_t>());
+      }
+   else
+#endif /* defined(JITSERVER_SUPPORT) */
+      {
+      TR::VMAccessCriticalSection dereferenceObjectPointerReferenceCriticalSection(self()->comp());
+      objectPointer = *objectReferenceLocation;
+      }
+   return objectPointer;
+   }
+
+
+TR::KnownObjectTable::Index
+J9::KnownObjectTable::getIndexAt(uintptrj_t *objectReferenceLocation)
+   {
+   TR::KnownObjectTable::Index result = UNKNOWN;
+#if defined(JITSERVER_SUPPORT)
+   if (self()->comp()->isOutOfProcessCompilation())
+      {
+      uintptrj_t objectPointer = self()->dereferenceObjectPointerReference(objectReferenceLocation);
```

Comment:
This implementation will send two messages: one for dereferencing the JNI reference given as parameter and another to do the search through the KOT. Maybe just send a message that will do both.